MoreLinq.Test/FromTest.cs

```diff
@@ -0,0 +1,94 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+
+namespace MoreLinq.Test
+{
+    using System;
+    using System.Collections.Generic;
+    using NUnit.Framework;
+
+    class FromTest
+    {
+        [Test]
+        public void TestFromIsLazy()
+        {
+            var breakingFunc = BreakingFunc.Of<int>();
+            MoreEnumerable.From(breakingFunc);
+            MoreEnumerable.From(breakingFunc, breakingFunc);
+            MoreEnumerable.From(breakingFunc, breakingFunc, breakingFunc);
+            MoreEnumerable.From(breakingFunc, breakingFunc, breakingFunc, breakingFunc);
+        }
+
+        [TestCase(1)]
+        [TestCase(2)]
+        [TestCase(3)]
+        [TestCase(4)]
+        public void TestFromInvokesMethods(int numArgs)
+        {
+            int F1() => -2;
+            int F2() => 4;
+            int F3() => int.MaxValue;
+            int F4() => int.MinValue;
+
+            switch (numArgs)
+            {
+                case 1: MoreEnumerable.From(F1).AssertSequenceEqual(F1()); break;
+                case 2: MoreEnumerable.From(F1, F2).AssertSequenceEqual(F1(), F2()); break;
+                case 3: MoreEnumerable.From(F1, F2, F3).AssertSequenceEqual(F1(), F2(), F3()); break;
+                case 4: MoreEnumerable.From(F1, F2, F3, F4).AssertSequenceEqual(F1(), F2(), F3(), F4()); break;
+                default: throw new ArgumentOutOfRangeException(nameof(numArgs));
+            }
+        }
+
+        [TestCase(1)]
+        [TestCase(2)]
+        [TestCase(3)]
+        [TestCase(4)]
+        public void TestFromInvokesMethodsMultipleTimes(int numArgs)
+        {
+            var evals = new [] { 0, 0, 0, 0 };
+            int F1() { evals[0]++; return -2; }
+            int F2() { evals[1]++; return -2; }
+            int F3() { evals[2]++; return -2; }
+            int F4() { evals[3]++; return -2; }
+
+            IEnumerable<int> results;
+            switch (numArgs)
+            {
+                case 1: results = MoreEnumerable.From(F1); break;
+                case 2: results = MoreEnumerable.From(F1, F2); break;
+                case 3: results = MoreEnumerable.From(F1, F2, F3); break;
+                case 4: results = MoreEnumerable.From(F1, F2, F3, F4); break;
+                default: throw new ArgumentOutOfRangeException(nameof(numArgs));
+            }
+
+            results.Consume();
+            results.Consume();
+            results.Consume();
+
+            for (int i = 0; i < numArgs; i++)
```

Comment:
Funny that I found the loop harder to read during the review. It slowed me down and I had to pause to mentally execute and make sure that all the moving parts of the loop are correct.

Setting aside our subjective views of the readability of each version, my (hopefully) objective argument would be that using LINQ makes the test safer and less prone to accidental edits. For example, if someone accidentally changed the `[i]` in `Assert.That(evals[i], Is.EqualTo(3))` to a `[0]` (reading `Assert.That(evals[0], Is.EqualTo(3))`)  then the test would still pass and yield a false positive. With the LINQ version, it would be a lot harder.

A loop inside the actual implementation (i.e. not the unit test) would be more acceptable for optimisation purposes.